### PR TITLE
Feature/create llm class

### PR DESCRIPTION
**Pull request by harryBedfordSL**
## Description

This PR refactors the `llm.py` file into its own module, introducing an abstract class for all models. We now have one implementation of this class, `mistral.py` , but we could now easily make an openai implementation and then swap out the model via config (there is a factory method that selects which implementation to use based on the type passed in).

Another important addition is the counting mechanism. The abstract llm class is making use of a metaclass, which will be used for all child implementations. In this class we are decorating the chat method with a counter, which increments and logs every time chat is called. The child implementations do not need to do anything extra, this is hidden from them.

## Changelog
- creation of llm abstract class
- create counting mechanism
- create mistral ai implementation of llm
- create mechanism (factory + config) of choosing what llm to use for what agent
- added tests
- changed supervisor return type to (Now None but could throw)

### CODE REVIEW

File: backend/src/supervisors/supervisor.py [hunk 1]
```diff
@@ -1,39 +1,39 @@
...
 from src.router import get_agent_for_task
 from src.agents import validator_agent
 
-no_tasks_response = "No tasks found to solve"
+no_questions_response = "No questions found to solve"
 unsolvable_response = "I am sorry, but I was unable to find an answer to this task"
 no_agent_response = "I am sorry, but I was unable to find an agent to solve this task"
 
+config = Config()
+router_llm = get_llm(config.router_llm)
```
Comment (hsauve-scottlogic) [inline backend/src/supervisors/supervisor.py:13]: Just wondering, shouldn't the supervisor be llm ignorant?

File: backend/src/agents/__init__.py [hunk 2]
```diff
@@ -6,16 +8,19 @@
 from .validator_agent import ValidatorAgent
 from .answer_agent import AnswerAgent
 
-validator_agent = ValidatorAgent()
-intent_agent = IntentAgent()
-answer_agent = AnswerAgent()
+config = Config()
+
+
+validator_agent = ValidatorAgent(get_llm(config.validator_agent_llm))
```
[the remark this answers is not in the document]
Comment (harryBedfordSL) [reply backend/src/agents/__init__.py:14]: Nope, just me overlooking that! yeah like it, will change

File: backend/src/llm/factory.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+from typing import Dict
+from .llm import LLM
+from .mistral import Mistral
+
+
+llm_map: Dict[str, LLM] = {"mistral": Mistral()}
```
[the remark this answers is not in the document]
Comment (harryBedfordSL) [reply backend/src/llm/factory.py:6]: like it, will have a play and see what I can do

**Review by hsauve-scottlogic — COMMENTED** [review]
Great work Harry, I've learnt a lot about ABC classes, and it's great to see a factory implementation. Just a couple of comments

File: backend/src/llm/factory.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
+from typing import Dict
+from .llm import LLM
+from .mistral import Mistral
+
+
+llm_map: Dict[str, LLM] = {"mistral": Mistral()}
+
+
+def get_llm(type: str | None) -> LLM:
+    if type is None:
```
[the remark this answers is not in the document]
Comment (harryBedfordSL) [reply backend/src/llm/factory.py:10]: That's right! I'm going to change it to `name` to help

File: backend/src/supervisors/supervisor.py (hunk 1 above)
Comment (harryBedfordSL) [reply backend/src/supervisors/supervisor.py:13]: it is a fair point, I'll move it down to the router file itself

File: backend/src/llm/factory.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (harryBedfordSL) [reply backend/src/llm/factory.py]: using `__init__` in llm.py with a class method, is working without the map

**Review by hsauve-scottlogic — APPROVED** [review]

**Review by hsauve-scottlogic — APPROVED** [review]

**Review by philbaxter — APPROVED** [review]

**Pull request merged**